File: 공통문제/4주차/다리를 지나는 트럭/PKH.py
def solution(bridge_length, weight, truck_weights):
    answer = 0
    temp1 = []; temp2 = []
    
    while True:
        # 대기 중인 트럭이 있을 경우
        if truck_weights:
            a = truck_weights.pop(0)
            temp1.append(a); temp2.append(0)
        
        # 운행 중인 트럭의 무게가 디리의 부하 무게보다 클 경우
        if sum(temp1) > weight:
            del temp1[-1]; del temp2[-1]
            truck_weights.insert(0,a)
        
        temp2 = [i+1 for i in temp2]
        if temp2[0] >= bridge_length:
            del temp1[0]; del temp2[0] # 맨 앞 차량이 다리를 다 건넌 것이기 떄문에 삭제
        answer += 1 # 시간은 계속 흘러가므로 +1
        
        # 아무것도 없을 때까지는 해야하니 시간에 +1
        if not temp1 and not truck_weights:
            answer += 1
            break
    return answer

File: 공통문제/4주차/다리를 지나는 트럭/PKH.java
import java.util.Deque;
import java.util.Iterator;
import java.util.ArrayDeque;

class Solution {
    public int solution(int bridge_length, int weight, int[] truck_weights) {
        int answer = 0; int time = 0;
        Deque<Integer> queue = makeDeque(truck_weights);
        Deque<Integer> onWeight = new ArrayDeque<Integer>();
        Deque<Integer> onLength = new ArrayDeque<Integer>();
        
        while(true) {
            int nowTruck = 0;
            if (queue.size() > 0) {
                nowTruck = queue.poll();
                onWeight.add(nowTruck);
                onLength.add(0);
            }
            
            if (sum(onWeight) > weight) {
                onWeight.removeLast(); onLength.removeLast();
                queue.addFirst(nowTruck);
            }
            
            for(int i = 0; i < onLength.size(); i++) {
                int temp = onLength.poll();
                temp += 1;
                onLength.addLast(temp);
            }
            
            if (onLength.getFirst() >= bridge_length) {
                onWeight.removeFirst(); onLength.removeFirst();
            }
            
            answer += 1;
            
            if (onWeight.size() == 0 && queue.size() == 0) {
                answer += 1;
                break;
            }
        }
        return answer;
    }
    
    public static Deque<Integer> makeDeque(int[] nums) {
        Deque<Integer> queue = new ArrayDeque<Integer>();
        for(int i : nums) {
            queue.add(i);
        }
        return queue;
    }
    
    public static int sum(Deque<Integer> deq) {
        int total = 0;
        Iterator<Integer> i = deq.iterator();
        while(i.hasNext()) {
            total += i.next();
        }
        return total;
    }
}
